src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+            => Task.FromResult(delegatedResponse);
+
+        /// <summary>
+        /// Extracts information about the request and then calls <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>.
+        /// Children can override to provide any more custom logic needed before doing a normal delegated response pass with <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>
+        /// </summary>
+        public virtual async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+            if (documentContext is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await _documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
+
+            return await GetDelegatedResponseAsync(request, documentContext, codeDocument, sourceText, absoluteIndex, projection, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected async Task<TResponse?> GetDelegatedResponseAsync(
+            TRequest request,
+            DocumentContext documentContext,
+            RazorCodeDocument codeDocument,
+            SourceText sourceText,
+            int absoluteIndex,
+            Projection projection,
+            CancellationToken cancellationToken)
+        {
+            if (!_languageServerFeatureOptions.SingleServerSupport)
+            {
+                return await HandleWithoutSingleServerAsync(request, documentContext, codeDocument, sourceText, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (projection.LanguageKind == RazorLanguageKind.Razor)
+            {
+                return await HandleInRazorAsync(request, documentContext, codeDocument, sourceText, cancellationToken).ConfigureAwait(false);
+            }
+
+            var delegatedParams = CreateDelegatedParams(request, documentContext, projection);
+
+            var delegatedRequest = await _languageServer.SendRequestAsync(EndpointName, delegatedParams).ConfigureAwait(false);
```

Comment:
I wonder if EndpointName should be renamed here.. given that there are a few end points in the mix, this is perhaps better called "CustomMessageTargetEndpointName" or something? Though thats a bit wordy I admit :P

Reply:
I was hoping that documentation on the property helped here. It's a lot going on, so a name probably eventually won't give enough information here. I'm not opposed to renaming, but unless there's an obvious switch I think keeping the documentation clear is more important.

Reply:
Fine by me, but I'd expand the doc comment a bit then, perhaps even to mention RazorCustomMessageTarget explicitly. All of the method names for the custom message target might be in LanguageServerConstants, but I don't think the reverse would be implied.

Reply:
No argument from me

Reply:
Done... it will make the review a bit messier but it's not too bad